rasterio/_vsiopener.pyx

```diff
@@ -77,46 +83,141 @@ cdef void uninstall_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callback
     callbacks_struct = NULL
 
 
-cdef void* pyopener_open(void *pUserData, const char *pszFilename, const char *pszAccess) except NULL with gil:
+cdef int pyopener_stat(
+    void *pUserData,
+    const char *pszFilename,
+    VSIStatBufL *pStatBuf,
+    int nFlags
+) except -1 with gil:
+    """Provides POSIX stat data to GDAL from a Python filesystem."""
+    # Convert the given filename to a registry key.
+    # Reminder: openers are registered by URI scheme, authority, and 
+    # *directory* path.
+    urlpath = pszFilename.decode("utf-8")
+    parsed_uri = urlparse(urlpath)
+    parent = Path(parsed_uri.path).parent
+
+    # Note that "r" mode is used here under the assumption that GDAL
+    # doesn't read_dir when writing data. Could be wrong!
+    mode = "r"
+    key = ((parsed_uri.scheme, parsed_uri.netloc, parent.as_posix()), mode)
+
+    registry = _OPENER_REGISTRY.get()
+    log.debug("Looking up opener in pyopener_stat: registry=%r, key=%r", registry, key)
+    try:
+        file_opener = registry[key]
+    except KeyError as err:
+        errmsg = f"Opener not found: {repr(err)}".encode("utf-8")
+        CPLError(CE_Failure, <CPLErrorNum>4, <const char *>"%s", <const char *>errmsg)
+        return -1
+
+    try:
+        if file_opener.isfile(urlpath):
+            fmode = 0o170000 | stat.S_IFREG
+        elif file_opener.isdir(urlpath):
+            fmode = 0o170000 | stat.S_IFDIR
+        else:
+            # No such file or directory.
+            return 1
+        size = file_opener.size(urlpath)
+        mtime = file_opener.mtime(urlpath)
+    except (FileNotFoundError, KeyError):
+        # No such file or directory.
+        return 1
```

Comment:
-1 is also expected for a non existing file:

```suggestion
        return -1
```

Reply:
@rouault it seemed to me that GDAL expects 1 for a failure to stat. -1 is a message from Cython to the Python interpreter to call PyErr_SetString (or another method) and raise an exception.

Reply:
> it seemed to me that GDAL expects 1 for a failure to stat.

no, it just does follow Unix stat() (https://linux.die.net/man/2/stat) : "On success, zero is returned. On error, -1 is returned, [and errno is set appropriately]. ".   I commented out the part about errno, because it just doesn't look at it.  That said I'd suspect most code in GDAL to just test VSIStatL() result against 0, and not test against -1, but it is probably be safer to follow the expected return codes

Reply:
@rouault oh, I see! Thank you.

I've also reminded myself that these callbacks have a different problem. The `except -1` or `except NULL` clauses can't have their intended effects. When GDAL calls these functions, there is no Python context, and so Cython can't  cause an exception to be raised. The best we can do is to use `CPLError()` and expect Rasterio to pick up that error when the callback's caller fails.

I've got more work to do here.